azurelinuxagent/ga/update.py

```diff
@@ -459,29 +456,64 @@ def __goal_state_updated(self, incarnation):
         return incarnation != self.last_incarnation
 
     def _process_goal_state(self, exthandlers_handler, remote_access_handler):
+
+        def log_next_update_time():
+            next_normal_time, next_hotfix_time = self.__get_next_upgrade_times()
+            upgrade_type = self.__get_agent_upgrade_type(available_agent)
+            next_time = next_hotfix_time if upgrade_type == AgentUpgradeType.Hotfix else next_normal_time
+            message_ = "Discovered new {0} upgrade {1}; Will upgrade on or after {2}".format(
+                upgrade_type, available_agent.name,
+                datetime.utcfromtimestamp(next_time).strftime(logger.Logger.LogTimeFormatInUTC))
+            add_event(AGENT_NAME, op=WALAEventOperation.AgentUpgrade, version=CURRENT_VERSION, is_success=True,
+                      message=message_, log_event=False)
+            logger.info(message_)
+
+        def handle_updates_for_requested_version():
+            if requested_version < CURRENT_VERSION:
+                prefix = "downgrade"
+                # In case of a downgrade, we blacklist the current agent to avoid starting it back up ever again
+                # (the expectation here being that if RSM is asking us to a downgrade,
+                # there's a good reason for not wanting the current version).
+                try:
+                    # We should always have an agent directory for the CURRENT_VERSION
+                    # (unless the CURRENT_VERSION == daemon version, but since we don't support downgrading
+                    # below daemon version, we will never reach this code path if that's the scenario)
+                    current_agent = next(agent for agent in self.agents if agent.version == CURRENT_VERSION)
+                    current_agent.mark_failure(is_fatal=True)
```

Comment:
Done (though I did mention the word `blacklisting` because I feel its pretty self explanatory (and makes it easier to explain) even to folks outside our dev group who read the logs for debugging.